### PR TITLE
Poseidon hs upgrade

### PR DESCRIPTION
**Pull request by stschiff**
OK, I have made all the changes necessary to work with the latest poseidon-hs. I am not yet ready quite yet, as I would like to also implement the stats block output. But I wanted to make this available for your review @nevrome, whenever you have time, since there are also changes to AdmixPop which need to be reviewed by you.

### CODE REVIEW

**Comment by nevrome** [comment]
This doesn't compile for me:

```
[ 5 of 11] Compiling Poseidon.Analysis.CLI.FStats

/home/schmid/agora/poseidon-analysis-hs/src/Poseidon/Analysis/CLI/FStats.hs:196:34: error:
    Variable not in scope: popLefts :: [a0]
    |
196 |                 forM_ (zip [0..] popLefts) $ \(i, popLeft) ->
    |                                  ^^^^^^^^

/home/schmid/agora/poseidon-analysis-hs/src/Poseidon/Analysis/CLI/FStats.hs:197:38: error:
    Variable not in scope: popRights :: [a1]
    |
197 |                     forM_ (zip [0..] popRights) $ \(j, popRight) ->
    |                                      ^^^^^^^^^

/home/schmid/agora/poseidon-analysis-hs/src/Poseidon/Analysis/CLI/FStats.hs:198:51: error:
    • Variable not in scope: blockData :: [BlockData]
    • Perhaps you meant data constructor ‘BlockData’ (line 84)
    |
198 |                         forM_ (zip [(0 :: Int)..] blockData) $ \(k, block) ->
    |                                                   ^^^^^^^^^

/home/schmid/agora/poseidon-analysis-hs/src/Poseidon/Analysis/CLI/FStats.hs:201:108: error:
    Variable not in scope: blockVals :: BlockData -> [[a3]]
    |
201 |                                 show (snd . blockEndPos $ block), show (blockSiteCount block !! i), show ((blockVals block !! i) !! j)]
    |                                                                                                            ^^^^^^^^^
```

I'll look into the changes now without running anything.

File: manual_tests/fstats_tests/runFstats.sh [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
+
+# .-----------.-----------.---------------.--------.-------.---------.------------.-----------.---------------------.
+# | Statistic |     a     |       b       |   c    |   d   | NrSites |  Estimate  |  StdErr   |       Z score       |
+# :===========:===========:===============:========:=======:=========:============:===========:=====================:
+# | F3        | French    | Italian_North | Mbuti  |       | 593124  | 0.2576     | 2.6932e-3 | 95.64402147158148   |
+# | F3        | French    | Han           | Mbuti  |       | 593124  | 0.2168     | 2.5383e-3 | 85.39316355835682   |
+# | F3        | Sardinian | Pima          | French |       | 593124  | -5.3477e-3 | 3.9419e-4 | -13.566531670301005 |
+# | F4        | French    | Russian       | Han    | Mbuti | 593124  | -1.6778e-3 | 9.1419e-5 | -18.35262346091248  |
+# | F4        | Sardinian | French        | Pima   | Mbuti | 593124  | -1.4384e-3 | 1.1525e-4 | -12.481084899924868 |
+# '-----------'-----------'---------------'--------'-------'---------'------------'-----------'---------------------'
```
Comment (nevrome) [inline manual_tests/fstats_tests/runFstats.sh:13]: Beautiful!

File: src/Poseidon/Analysis/CLI/FStats.hs [hunk 2]
```diff
@@ -96,12 +96,12 @@ data BlockAccumulator = BlockAccumulator
     , accMaybeEndPos   :: IORef (Maybe GenomPos)
     , accCount         :: IORef Int
     -- this is the key value accumulator: for each statistic there is a list of accumulators if needed, see above.
-    , accValues        :: V.Vector (VUM.IOVector Double) -- this is the key value accumulator: for each statistic there is a list of accumulators if needed, see above.
+    , accValues        :: V.Vector (VUM.IOVector Double)
     -- the outer vector is boxed and immutable, the inner vector is unboxed and mutable, so that values can be updated as we loop through the data.
```
Comment (nevrome) [inline src/Poseidon/Analysis/CLI/FStats.hs:100]: Clever!

File: src/Poseidon/Analysis/CLI/FStats.hs [hunk 3]
```diff
@@ -163,12 +173,33 @@ runFstats opts = do
...
             Just outFn -> liftIO . withFile outFn WriteMode $
                 \h -> mapM_ (hPutStrLn h . intercalate "\t") (tableH : tableB)
+
+        -- optionally output the block data
+        case _foBlockTableOut opts of
+            Nothing -> return ()
+            Just fn -> liftIO . withFile fn WriteMode $ \h -> do
+                let headerLine = if hasAscertainment 
+                     then ["Statistic", "a", "b", "c", "d", "NrSites", "Asc (Og, Ref)", "Asc (Lo, Up)", "Estimate", "StdErr", "Z score"]
+                     else ["Statistic", "a", "b", "c", "d", "NrSites", "Estimate", "StdErr", "Z score"]
```
Comment (nevrome) [inline src/Poseidon/Analysis/CLI/FStats.hs:194]: Already in line 165-167.

File: src/Poseidon/Analysis/Utils.hs [hunk 4]
```diff
@@ -33,12 +32,18 @@ type GenomPos = (Chrom, Int)
 
 type GroupDef = (String, SignedEntitiesList)
 
+-- redundant non-exported copy from Poseidon.EntitiesList
+meansIn :: SelectionLevel2 -> Bool
+meansIn ShouldBeIncluded                   = True
+meansIn ShouldBeIncludedWithHigherPriority = True
+meansIn ShouldNotBeIncluded                = False
```
Comment (nevrome) [inline src/Poseidon/Analysis/Utils.hs:39]: Right - you already mentioned this. Can be removed in the future mono-repo.

File: src/Poseidon/Generator/CLI/AdmixPops.hs [hunk 5]
```diff
@@ -32,19 +33,19 @@ data AdmixPopsOptions = AdmixPopsOptions {
...
-    , _readOptStopOnDuplicates = False
+      _readOptStopOnDuplicates = False
     , _readOptIgnoreChecksums  = True
     , _readOptIgnoreGeno       = False
     , _readOptGenoCheck        = True
     }
 
-runAdmixPops :: AdmixPopsOptions -> PoseidonLogIO ()
-runAdmixPops (AdmixPopsOptions genoSources popsWithFracsDirect popsWithFracsFile marginalizeMissing outFormat outPath maybeOutName) = do
+runAdmixPops :: AdmixPopsOptions -> PoseidonIO ()
```
Comment (nevrome) [inline src/Poseidon/Generator/CLI/AdmixPops.hs:47]: The changes here are minimal. Looks all fine to me! Thanks!

**Review by nevrome — APPROVED** [review]
All good as far as I can judge just by staring at the code. Generally I think it's better to do separate refactoring and implementing new features, but I understand that sometimes it can be convenient to do everything in one go.

**Comment by stschiff** [comment]
Yes, I'm sorry it doesn't compile. I started adding stuff after I started the PR, my bad. Thanks for reviewing the bits you did. I'll finalize now and make ready for review/merge.

**Comment by stschiff** [comment]
OK, I've followed your advice and removed the block output stuff from this branch for now. If you give green-light, I will merge this and then make another PR with the block-output feature.

**Comment by nevrome** [comment]
Excellent! Can be merged now :+1:

**Comment by nevrome** [comment]
I think this PR will close #8 and #10, right?

**Pull request merged**